gdxsv/lbs_discord.go

```diff
@@ -0,0 +1,513 @@
+					Value: v,
+				})
+				accumulatedEmbedStringLength += len(name) + len(v)
+			}
+
+		}
+		for i, fields := range splitEmbedFields(lobbyFields) {
+			description := ""
+			if i == 0 {
+				description = fmt.Sprintf("🌐 **待機中 %d 人**", lobbyUserCount)
+			}
+			payload.Embed = append(payload.Embed, &discordEmbed{
+				Description: description,
+				Color:       24041,
+				Fields:      fields,
+			})
+			accumulatedEmbedStringLength += len(description)
+		}
+
+		//
+		//3rd Embed, battle count
+		//
+		var battleFields []*discordEmbedField
+		for _, b := range battle {
+
+			value := addBlankIfRequired(b.RenpoUsersString + b.ZeonUsersString)
+
+			for i, v := range splitEmbedFieldValues(value) {
+				name := "⠀"
+				if i == 0 {
+					name = b.RegionName
+				}
+
+				battleFields = append(battleFields, &discordEmbedField{
+					Name:  name,
+					Value: v,
+				})
+				accumulatedEmbedStringLength += len(name) + len(v)
+			}
+		}
+		for i, fields := range splitEmbedFields(battleFields) {
+			description := ""
+			if i == 0 {
+				description = fmt.Sprintf("💥 **戦闘中 %d 人**", battleUserCount)
+			}
+			payload.Embed = append(payload.Embed, &discordEmbed{
+				Description: description,
+				Color:       13179394,
+				Fields:      fields,
+			})
+			accumulatedEmbedStringLength += len(description)
+		}
+
+		//
+		// Create the json
+		//
+		buffer := &bytes.Buffer{}
+		encoder := json.NewEncoder(buffer)
+		encoder.SetEscapeHTML(false)
+		err := encoder.Encode(payload)
+
+		if err != nil {
+			logger.Error("Failed to create Discord JSON", zap.Error(err))
+			return
+		}
+
+		//embed structure must not exceed 6000 characters
+		if accumulatedEmbedStringLength > 6000 {
+			originalSize := buffer.Len()
+
+			start := time.Now()
+			jsonString := reduceJSONStringSize(buffer.String())
+			buffer.Reset()
+			buffer.Write([]byte(jsonString))
+			elapsed := time.Since(start)
+
+			logger.Info("Reduce JSON Size!", zap.Int("original size", originalSize), zap.Any("elapsed", elapsed))
+		}
+
+		logger.Info(buffer.String(), zap.Int("length", buffer.Len()))
+
+		//
+		// Send to Discord
+		//
+		send := func(buf *bytes.Buffer) {
+
+			req, err := http.NewRequest("PATCH", conf.DiscordLiveStatusWebhookURL, buf)
+			req.Header.Set("Content-Type", "application/json")
+
+			resp, err := http.DefaultClient.Do(req)
+			if err != nil {
+				logger.Error("Failed to publish to Discord", zap.Error(err))
+				return
+			}
+			defer resp.Body.Close()
+
+			if ratelimitRemaining := resp.Header.Get("x-ratelimit-remaining"); len(ratelimitRemaining) > 0 {
+				discordLiveStatusRateRemaining, _ = strconv.Atoi(ratelimitRemaining)
+			}
+			if ratelimitReset := resp.Header.Get("x-ratelimit-reset"); len(ratelimitReset) > 0 {
+				discordLiveStatusRateResetEpoch, _ = strconv.ParseInt(ratelimitReset, 10, 64)
+			}
+
+			logger.Info("Discord Webhook sent", zap.String("Status", resp.Status))
+			if resp.StatusCode == http.StatusBadRequest {
+				body, err := ioutil.ReadAll(resp.Body)
```

Comment:
We must read `resp.Body` regardless of status code.
If there is data left in the buffer, the connection will not be reused.

Reply:
Oh I thought only `defer resp.Body.Close()` is required, thanks